compiler/passes/denormalize.cpp

```diff
@@ -0,0 +1,602 @@
+
+  collectSymbolSetSymExprVec(fn, symSet, symExprs);
+  buildDefUseMaps(symSet, symExprs, defMap, useMap);
+
+  //data structures for deferred actual handling
+  std::set<CallExpr*> deferredFns;
+  ActualUseDefCastMap actualUseDefMap;
+
+  bool cachedGlobalManip = globalManipFuncCache.count(fn);
+
+  forv_Vec(Symbol, sym, symSet) {
+
+    SymExpr *use = NULL;
+    Expr *usePar = NULL;
+    Expr *def = NULL;
+    Expr *defPar;
+    Type* castTo = NULL;
+
+    //if we don't already have it cached,
+    //check for global symbols in function body
+    if(!cachedGlobalManip) {
+      //I am not sure if there should be !isConstant and !isParam as well
+      //I think there can be a const but stateful global that is manipulated by
+      //the function so, play safe
+      if(sym && !sym->isImmediate() && isGlobal(sym)){
+        globalManipFuncCache[fn] = true;
+        cachedGlobalManip = true;
+      }
+    }
+
+    if(isDenormalizable(sym, defMap, useMap, &use, &def, &castTo)) {
+      usePar = use->parentExpr;
+      defPar = def->parentExpr;
+
+      //defer if the symbol used as actual
+      if(CallExpr* useCe = toCallExpr(usePar)){
+        if(!useCe->isPrimitive()) {
+          deferredFns.insert(useCe);
+          std::pair<Expr*, Type*> defCastBundle(def, castTo);
+          actualUseDefMap.insert(std::pair<SymExpr*, std::pair<Expr*, Type*> >
+              (use, defCastBundle));
+          continue;
+        }
+      }
+
+      //denormalize if the def is safe to move and there is no unsafe function
+      //between use and def
+      if(isExprSafeForReorder(def)) {
+        if(!possibleDepInBetween(defPar, usePar)) {
+          denormalize(def, use, castTo);
+        }
+      }
+    }
+  } // end loop for symbol
+
+  if(!cachedGlobalManip) {
+    globalManipFuncCache[fn] = false;
+  }
+
+  //handle deferred actuals
+  for(std::set<CallExpr*>::iterator ceIt = deferredFns.begin() ;
+      ceIt != deferredFns.end() ; ceIt++) {
+    denormalizeActuals(*ceIt, defMap, useMap, actualUseDefMap);
+  }
+
+  freeDefUseMaps(defMap, useMap);
+}
+
+void denormalizeActuals(CallExpr* ce,
+    Map<Symbol*,Vec<SymExpr*>*> defMap,
+    Map<Symbol*,Vec<SymExpr*>*> useMap,
+    ActualUseDefCastMap actualUseDefMap) {
+
+  INT_ASSERT(!ce->isPrimitive());
+  for_alist_backward(actual, ce->argList) {
+    if(SymExpr* argSym = toSymExpr(actual)) { //else it's already denormd
+      if(! (argSym->var->isConstant() || argSym->var->isParameter())) {
+        if(actualUseDefMap.count(argSym) > 0) {
+          std::pair<Expr*,Type*> tmpTuple = actualUseDefMap[argSym];
+          SymExpr* use = argSym;
+          Expr* def = tmpTuple.first;
+          Type* castTo = tmpTuple.second;
+          Expr* usePar = use->parentExpr;
+          Expr* defPar = def->parentExpr;
+          if(CallExpr* ceTmp = toCallExpr(defPar)) {
+            if(!isRecord(ceTmp->get(1)->typeInfo())) { //to preserve pass-by-value
+              if(isExprSafeForReorder(def)) {
```

Comment:
And I suppose modifications to state in between the def and the use are OK if you can prove that the def and the use couldn't read/write the same state.
Also note that any use of atomic or sync variable should count as "modifies state" to support the memory consistency model.